sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp

```diff
@@ -254,43 +263,51 @@ class annotated_ptr {
 template <typename T, typename... Props>
 class __SYCL_SPECIAL_CLASS
 __SYCL_TYPE(annotated_ptr) annotated_ptr<T, detail::properties_t<Props...>> {
-  using property_list_t = detail::properties_t<Props...>;
 
   static_assert(std::is_same_v<T, void> || std::is_trivially_copyable_v<T>,
                 "annotated_ptr can only encapsulate either "
                 "a trivially-copyable type "
                 "or void!");
 
-  // buffer_location and alignment are allowed for annotated_ref
-  // Cache controls are allowed for annotated_ptr
-  using allowed_properties =
-      std::tuple<decltype(ext::intel::experimental::buffer_location<0>),
-                 decltype(ext::oneapi::experimental::alignment<0>),
-                 decltype(ext::intel::experimental::read_hint<
-                          ext::intel::experimental::cache_control<
-                              ext::intel::experimental::cache_mode::cached,
-                              cache_level::L1>>),
-                 decltype(ext::intel::experimental::read_assertion<
-                          ext::intel::experimental::cache_control<
-                              ext::intel::experimental::cache_mode::cached,
-                              cache_level::L1>>),
-                 decltype(ext::intel::experimental::write_hint<
-                          ext::intel::experimental::cache_control<
-                              ext::intel::experimental::cache_mode::cached,
-                              cache_level::L1>>)>;
-  using filtered_properties =
-      typename PropertiesFilter<allowed_properties, Props...>::tuple;
-
-  // template unpack helper
-  template <typename... FilteredProps> struct unpack {};
+  using property_list_t = detail::properties_t<Props...>;
 
-  template <typename... FilteredProps>
-  struct unpack<std::tuple<FilteredProps...>> {
-    using type = detail::properties_t<FilteredProps...>;
+  // annotated_ref type
+  using reference =
+      sycl::ext::oneapi::experimental::annotated_ref<T, property_list_t>;
+
+  // operator enable/disable check
+  enum operator_id {
+    op_plus,
+    op_minus,
+    op_subscript,
+    op_inc,
+    op_dec,
   };
 
-  using reference = sycl::ext::oneapi::experimental::annotated_ref<
-      T, typename unpack<filtered_properties>::type>;
+#define OP_NOT_SUPPORTED(op, property)                                         \
+  "operator" op " is not available when " property " is specified!"
+
+  static constexpr void operatorAvailablityCheck(constexpr unsigned op_id) {
+    static constexpr bool hasAlign =
+        detail::ContainsProperty<alignment_key, std::tuple<Props...>>::value;
```

Comment:
this shouldn't use tuple either.

Reply:
This is the expected input from detail::ContainsProperty

Reply:
yeah but it shouldn't be. OK if you don't want to do that as part of this PR.

Reply:
I see. We should do this in another PR because there are other files need to be updated